x/staking/keeper/determinstic_test.go

```diff
@@ -0,0 +1,945 @@
+package keeper_test
+
+import (
+	"testing"
+	"time"
+
+	"cosmossdk.io/math"
+	"github.com/stretchr/testify/suite"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	"pgregory.net/rapid"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	"github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
+	simtestutil "github.com/cosmos/cosmos-sdk/testutil/sims"
+	"github.com/cosmos/cosmos-sdk/testutil/testdata"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	banktestutil "github.com/cosmos/cosmos-sdk/x/bank/testutil"
+	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
+	stakingkeeper "github.com/cosmos/cosmos-sdk/x/staking/keeper"
+	stakingtestutil "github.com/cosmos/cosmos-sdk/x/staking/testutil"
+
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+var (
+	validator1        = "cosmosvaloper1qqqryrs09ggeuqszqygqyqd2tgqmsqzewacjj7"
+	validatorAddr1, _ = sdk.ValAddressFromBech32(validator1)
+	validator2        = "cosmosvaloper1gghjut3ccd8ay0zduzj64hwre2fxs9ldmqhffj"
+	validatorAddr2, _ = sdk.ValAddressFromBech32(validator2)
+	delegator1        = "cosmos1nph3cfzk6trsmfxkeu943nvach5qw4vwstnvkl"
+	delegatorAddr1    = sdk.MustAccAddressFromBech32(delegator1)
+	delegator2        = "cosmos139f7kncmglres2nf3h4hc4tade85ekfr8sulz5"
+	delegatorAddr2    = sdk.MustAccAddressFromBech32(delegator2)
+)
+
+type DeterministicTestSuite struct {
+	suite.Suite
+
+	ctx           sdk.Context
+	stakingKeeper *stakingkeeper.Keeper
+	bankKeeper    bankkeeper.BaseKeeper
+	accountKeeper authkeeper.AccountKeeper
+	queryClient   stakingtypes.QueryClient
+	amt1          math.Int
+	amt2          math.Int
+}
+
+func (s *DeterministicTestSuite) SetupTest() {
+	var interfaceRegistry codectypes.InterfaceRegistry
+
+	app, err := simtestutil.Setup(
+		stakingtestutil.AppConfig,
+		&s.bankKeeper,
+		&s.accountKeeper,
+		&s.stakingKeeper,
+		&interfaceRegistry,
+	)
+	s.Require().NoError(err)
+
+	s.ctx = app.BaseApp.NewContext(false, tmproto.Header{Height: 1, Time: time.Now()}).WithGasMeter(sdk.NewInfiniteGasMeter())
+
+	queryHelper := baseapp.NewQueryServerTestHelper(s.ctx, interfaceRegistry)
+	stakingtypes.RegisterQueryServer(queryHelper, stakingkeeper.Querier{Keeper: s.stakingKeeper})
+	s.queryClient = stakingtypes.NewQueryClient(queryHelper)
+
+	s.amt1 = s.stakingKeeper.TokensFromConsensusPower(s.ctx, 101)
+	s.amt2 = s.stakingKeeper.TokensFromConsensusPower(s.ctx, 102)
+}
+
+func TestDeterministicTestSuite(t *testing.T) {
+	suite.Run(t, new(DeterministicTestSuite))
+}
+
+func drawDuration() *rapid.Generator[time.Duration] {
```

Comment:
```suggestion
func durationGenerator() *rapid.Generator[time.Duration] {
```

"draw" is used to pick one from a given generator